### PR TITLE
Feature/http nextgen refactor

### PR DESCRIPTION
**Pull request by graetzer**
### Scope & Purpose

Refactor code for future support of newever HTTP versions

- [x] Strictly *new functionality* (i.e. a new feature / new option, no need for porting)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

#### Related Information

- [x] There is an internal planning ticket: 
- [x] There is a *JIRA Ticket number* (In case a customer was affected / involved): https://arangodb.atlassian.net/browse/FEAT-201

### Testing & Verification

This change is already covered by existing tests, such as http_server, ssl_server

http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr/7954/

### CODE REVIEW

File: arangod/GeneralServer/AcceptorTcp.cpp [hunk 1]
```diff
@@ -189,17 +186,15 @@ void AcceptorTcp<SocketType::Ssl>::performHandshake(std::unique_ptr<AsioSocket<S
...
   });
 
   auto cb = [this, as = std::move(proto)](asio_ns::error_code const& ec) mutable {
     as->timer.cancel();
     if (ec) {
       LOG_TOPIC("4c6b4", DEBUG, arangodb::Logger::COMMUNICATION)
           << "error during TLS handshake: '" << ec.message() << "'";
-      asio_ns::error_code err;
-      as->shutdown(err);  // ignore error
+      as.reset(); // ungraceful shutdown
```
Comment (ObiWahn) [inline arangod/GeneralServer/AcceptorTcp.cpp]: https://www.boost.org/doc/libs/1_71_0/doc/html/boost_asio/reference/basic_stream_socket/close/overload2.html -- see Remarks.

Why is it ok to kill / reset the connection? Will there be no problem for clients (They could wait for a response on a connection that is not closed for them)?

File: arangod/GeneralServer/GeneralCommTask.cpp [hunk 2]
```diff
@@ -35,44 +35,72 @@ using namespace arangodb::rest;
...
     _protocol->timer.cancel();
-    asio_ns::error_code ec;
-    _protocol->shutdown(ec);
-    if (ec) {
-      LOG_TOPIC("2c6b4", DEBUG, arangodb::Logger::REQUESTS)
-          << "error shutting down asio socket: '" << ec.message() << "'";
-    }
+    LOG_DEVEL << "pre shutdown";
+    _protocol->shutdown([this, self(shared_from_this())](asio_ns::error_code ec) {
+      LOG_DEVEL << "post shutdown";
```
Comment (jsteemann) [inline arangod/GeneralServer/GeneralCommTask.cpp]: should be removed

File: arangod/GeneralServer/GeneralCommTask.cpp [hunk 3]
```diff
@@ -35,44 +35,72 @@ using namespace arangodb::rest;
...
+      if (ec) {
+        LOG_TOPIC("2c6b4", ERR, arangodb::Logger::REQUESTS)
```
Comment (jsteemann) [inline arangod/GeneralServer/GeneralCommTask.cpp]:
```suggestion
        LOG_TOPIC("2c6b4", WARN, arangodb::Logger::REQUESTS)
```

File: arangod/GeneralServer/GeneralServer.cpp [hunk 4]
```diff
@@ -120,19 +120,37 @@ void GeneralServer::startListening() {
...
 }
 
 void GeneralServer::stopWorking() {
   _acceptors.clear();
+  std::unique_lock<std::recursive_mutex> guard(_tasksLock);
+  auto now = std::chrono::system_clock::now();
+  while (!_commTasks.empty()) {  // CommTasks should deregister themselves
+    guard.unlock();
+    std::this_thread::yield();
+    if ((std::chrono::system_clock::now() - now) < std::chrono::seconds(3)) {
```
Comment (jsteemann) [inline arangod/GeneralServer/GeneralServer.cpp]: Shouldn't we check for if *more* than 3 seconds passed, not less?
```suggestion
    if ((std::chrono::system_clock::now() - now) > std::chrono::seconds(3)) {
```

File: arangod/GeneralServer/IoContext.cpp [hunk 5]
```diff
@@ -54,16 +54,17 @@ IoContext::IoContext(IoContext const& other)
...
 }
 
 IoContext::~IoContext() { stop(); }
 
 void IoContext::stop() {
+  _work.reset();
   io_context.stop();
   while(_thread.isRunning()) {
-    cpu_relax();
+    std::this_thread::yield();
```
Comment (ObiWahn) [inline arangod/GeneralServer/IoContext.cpp]: Linus will get mad :)

File: lib/Rest/HttpRequest.cpp [hunk 6]
```diff
@@ -158,25 +145,28 @@ void HttpRequest::parseHeader(char* start, size_t length) {
 
               // HTTP protocol version is expected next
               // trim value
+#if 0
```
Comment (jsteemann) [inline lib/Rest/HttpRequest.cpp]: still needed?

File: lib/Rest/HttpRequest.cpp [hunk 7]
```diff
@@ -711,12 +705,12 @@ void HttpRequest::setHeader(char const* key, size_t keyLength,
   TRI_ASSERT(key != nullptr);
   TRI_ASSERT(value != nullptr);
 
-  if (keyLength == StaticStrings::ContentLength.size() &&
-      memcmp(key, StaticStrings::ContentLength.c_str(), keyLength) == 0) {  // 14 = strlen("content-length")
-    _contentLength = NumberUtils::atoi_zero<int64_t>(value, value + valueLength);
-    // do not store this header
-    return;
-  }
+//  if (keyLength == StaticStrings::ContentLength.size() &&
```
Comment (jsteemann) [inline lib/Rest/HttpRequest.cpp]: still needed?

File: lib/Rest/HttpRequest.h [hunk 8]
```diff
@@ -113,20 +103,19 @@ class HttpRequest final : public GeneralRequest {
   void setArrayValue(char const* key, size_t length, char const* value);
 
  private:
+  
+  /// used by RestBatchHandler (an API straight from hell)
```
Comment (jsteemann) [inline lib/Rest/HttpRequest.h]: :+1:

File: arangod/GeneralServer/CommTask.cpp [hunk 9]
```diff
@@ -530,46 +571,41 @@ bool CommTask::handleRequestAsync(std::shared_ptr<RestHandler> handler,
...
-    result = rest::ResponseCode::UNAUTHORIZED;
+  // deny access to database with NONE, (except /_api/user)
+  if (vc->databaseAuthLevel() == auth::Level::NONE/* && !StringUtils::isPrefix(path, ApiUser)*/) {
+    result = Flow::Abort;
     LOG_TOPIC("0898a", TRACE, Logger::AUTHORIZATION) << "Access forbidden to " << path;
 
-    if (userAuthenticated) {
-      req.setAuthenticated(false);
-    }
+//    if (userAuthenticated) {  // hack to not confuse RestHandlers
```
Comment (jsteemann) [inline arangod/GeneralServer/CommTask.cpp]: still needed?

**Comment by ObiWahn** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-mac-onlycatch/5843/ - retry mac build

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-linux/9749/

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-linux/9753/

**Comment by graetzer** [comment]
http://jenkins01.arangodb.biz:8080/job/arangodb-matrix-pr-linux/9759/

**Pull request merged**